### PR TITLE
Skip using rendezvousTime with beam weapon turrets

### PR DESCRIPTION
**Pull request by kaol**
**Bugfix**

## Fix Details
`AimTurrets` uses `RendezvousTime` to calculate where the target would be when accounting for projectile velocity. This works as expected for projectile turrets but beam turrets don't actually have a velocity but the velocity they use is instead their range. This makes the turret's aim go wrong with small and fast targets, sometimes.

Instead they can use a `rendezvousTime` value of zero. There was a number of subsequent calculations that could be skipped if `rendezvousTime` was known to be zero.

Someone who knows better about weapons code could check this as well. Should this be using `Lifetime` instead of `TotalLifetime`? I looked at the code and there was something about submunitions and I don't know if this code should care about them if the goal is to just make an initial hit at a target.

## Testing Done
I flew around a bit and added a turreted projectile weapon and it seemed to aim well as well.

## Save File
Omitted, this can be tested anywhere where there are small ships and if you have turrets with beam weapons. The bug will occur when using 01196eb963a9b16161cac7cfff19c898eda8bc22, and will not occur when using this branch's build.

### CODE REVIEW

**Comment by Hurleveur** [comment]
Nice! I noticed this bug but I was not sure if it came from code or just ship confusion.

**Comment by kaol** [comment]
Right, that code is also using distance to penalize targets that are out of it. I changed the PR to make beam weapons to use to `rendezvousTime` also when the target is out of range. No need to use the `RendezvousTime` function for it since it's a simple distance check.

I think the `Weapon` class could use some `IsBeam` function that would return a boolean if it was meant to act as a one, but I'd leave that to some other refactoring PR.

File: source/AI.cpp [hunk 1]
```diff
@@ -2861,22 +2861,38 @@ void AI::AimTurrets(const Ship &ship, FireCommand &command, bool opportunistic)
...
+					double distance = p.Length();
+					if(distance > weapon->Velocity())
+						rendezvousTime = distance / weapon->Velocity() - 1.;
+					else
+						rendezvousTime = 0.;
+				}
+				else
+				{
+					// Find out how long it would take for this projectile to reach the target.
+					rendezvousTime = weapon->TotalLifetime() == 1 ? 0 : RendezvousTime(p, v, vp);
```
Comment (Amazinite) [inline source/AI.cpp:2878]: This bit will never run.
```suggestion
					rendezvousTime = RendezvousTime(p, v, vp);
```

File: source/AI.cpp [hunk 2]
```diff
@@ -2861,22 +2861,38 @@ void AI::AimTurrets(const Ship &ship, FireCommand &command, bool opportunistic)
...
 				p += v;
 
-				// Find out how long it would take for this projectile to reach the target.
-				double rendezvousTime = RendezvousTime(p, v, vp);
-				// If there is no intersection (i.e. the turret is not facing the target),
-				// consider this target "out-of-range" but still targetable.
-				if(std::isnan(rendezvousTime))
-					rendezvousTime = max(p.Length() / (vp ? vp : 1.), 2 * weapon->TotalLifetime());
+				double rendezvousTime;
+				// Beam weapons hit instantenously if they are in range.
```
Comment (tibetiroka) [inline source/AI.cpp:2865]:
```suggestion
				// Beam weapons hit instantaneously if they are in range.
```

File: source/AI.cpp [hunk 3]
```diff
@@ -2861,22 +2861,38 @@ void AI::AimTurrets(const Ship &ship, FireCommand &command, bool opportunistic)
...
 
-				// Determine where the target will be at that point.
-				p += v * rendezvousTime;
+					// Determine where the target will be at that point.
+					p += v * rendezvousTime;
+
+					// All bodies within weapons range have the same basic
+					// weight. Outside that range, give them lower priority.
+					rendezvousTime = max(0., rendezvousTime - weapon->TotalLifetime());
+
```
Comment (tibetiroka) [inline source/AI.cpp:2890]:
```suggestion
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]
Just to clarify: this is more of a grammar review, not a code review. The code patrol team will take care of that.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
This only fixes turreted beam weapons, I don't see why we wouldn't fix gun beam weapons in the same PR, which I'm pretty sure have the same "bug". (the ship will angle at the wrong directing with its guns, its most visible when it only uses beam weapons I think)

**Comment by Hurleveur** [comment]
If you need any help doing something never hesitate to ask, here or in the discord!

**Comment by Hurleveur** [comment]
Nvm for the guns lets just fix it for turrets in this PR.

**Review by Hurleveur — APPROVED** [review]
Looks like a good fix.

**Pull request merged**